sklego/loess.py

```diff
@@ -0,0 +1,332 @@
+
+        x_focal = np.asarray([x_focal]).reshape(-1, 1)
+
+        x_low_indices = np.where(low_mask)[0]
+        x_high_indices = np.where(high_mask)[0]
+
+        if (x_low.size > floor(n_points / 2)) & (x_high.size > floor(n_points / 2)):
+            knn_low = NearestNeighbors(n_neighbors=int(n_points / 2)).fit(x_low.reshape(-1, 1))
+            knn_high = NearestNeighbors(n_neighbors=int(n_points / 2)).fit(x_high.reshape(-1, 1))
+            indices_low_local = knn_low.kneighbors(x_focal)[1][0]
+            indices_high_local = knn_high.kneighbors(x_focal)[1][0]
+
+            indices_low = x_low_indices[indices_low_local]
+            indices_high = x_high_indices[indices_high_local]
+            self.indices[index] = np.concatenate([indices_low, indices_high])
+
+        elif (x_low.size > floor(n_points / 2)) & (not x_high.size > floor(n_points / 2)):
+            knn_low = NearestNeighbors(n_neighbors=int(n_points)).fit(x_low.reshape(-1, 1))
+            indices_low_local = knn_low.kneighbors(x_focal)[1][0]
+            self.indices[index] = x_low_indices[indices_low_local]
+
+        else:
+            knn_high = NearestNeighbors(n_neighbors=int(n_points)).fit(x_high.reshape(-1, 1))
+            indices_high_local = knn_high.kneighbors(x_focal)[1][0]
+            self.indices[index] = x_high_indices[indices_high_local]
+
+    def _fit_model_per_window(self, x, y):
+        """
+        With the created x focal points and accompanying windows, fit per x focal point a model on
+        the local window and predict an y focal using the fitted model.
+        :param x:
+        :param y:
+        """
+        for index in self.indices.keys():
+            x_focal = np.asarray([self.x_focal_base[index]]).reshape(-1, 1)
+            x_window = x[self.indices[index]].reshape(-1, 1)
+            y_window = y[self.indices[index]].reshape(-1, 1)
+
+            # TODO: implement iterative weighted fit
+            model = self.model.fit(x_window, y_window)
+            self.y_focal_base = np.append(self.y_focal_base, model.predict(x_focal))
+            self.model_per_window = np.append(self.model_per_window, model)
+
+
+def plot_windows(x, y, x_focal_base, y_focal_base, indices):
+    """
+    TODO: Use matplotlib.animation to create video
+    Plot the data used in each window and it's accompanying focal point. Useful for debugging
+    methods used in the self._get_window_indices function
+    :param x:
+    :param y:
+    :param x_focal_base:
+    :param y_focal_base:
+    :param indices:
+    """
+    for index in indices.keys():
+        ax = plt.axes(xlim=(np.min(x), np.max(x)), ylim=(np.min(y), np.max(y)))
+        ax.scatter(x[indices[index]], y[indices[index]])
+        ax.scatter(x_focal_base[index], y_focal_base[index], c='r')
+        ax.set_title(f'Window for {x_focal_base[index]}')
+        plt.pause(0.01)
+        plt.show()
+
+
+# HELPER FUNCTIONS TO GENERATE RANDOM DATA
+def random_x(minimum_val, maximum_val, size):
```

Comment:
this should be part of `sklego.datasets` if we think it is important enough to have.

Reply:
I guess the functions itself are not that important, will move them to the development notebook.